### ISSUE TITLE
Coroutine[Any, Any, None] is not awaitable

### ISSUE THREAD

**Issue by failable** [microsoft/pyright#9204]
Version: `1.1.384`

![Image](https://github.com/user-attachments/assets/f394156d-a265-4a43-a53e-05c5adadee4d)

```
import asyncio


async def test() -> int:
    await asyncio.sleep(1)
    print("Good")
    return 1


async def main():
    await test()


if __name__ == "__main__":
    asyncio.run(main())

```

**Comment by tino** [comment]
I have the same issue, also in Zed. However I don't have it when I run my homebrew installed pyright on the same code, so might be something with the version in Zed?

**Comment by KimigaiiWuyi** [comment]
I have the exact same problem.
And I've noticed that these kinds of errors can randomly appear or randomly disappear without any potential change, sometimes by doing nothing, after a while he'll show up with an error, and similarly, even by doing nothing, after a while it won't be flagged as an error

This could be a performance issue?

![Image](https://github.com/user-attachments/assets/ec551fb5-8e6e-48d0-ac53-264e19501963)

**Comment by BaconPancakes** [comment]
Also reporting a similar issue. Either all the `await` keywords will return an error or None. ~~It is intermittent at times. I also can't reproduce in pyright sandbox nor in a fresh .venv, unfortunately. I suspect it has to do with project size.~~

I figured out something related, but it is still intermittent in terms of reproducibility. If you use `nest_asyncio` in a notebook (*.ipnyb), then the errors can start to appear. 

```
import nest_asyncio

nest_asyncio.apply()

On my main project when I remove `nest_asyncio`, the errors disappear across the project. I'm not sure if the notebook has anything to do with, I was just trying to reproduce the error from my main project.


Python version 3.12.5, 
Pylance language server 2024.10.100 (pyright version 1.1.384, commit f2dcf075)

![Image](https://github.com/user-attachments/assets/bb00ea5a-f52e-40f9-9741-e548d02650bd)

**Comment by erictraut** [comment]
Thanks for the additional clues. These are helpful, so keep them coming!

Is this behavior something that started recently? If so, that would point to a recent code change as the cause.

If you restart the language server, does the problem go away?

Are you able to repro the problem with the command-line version of pyright?

Does anyone have a consistent repro, or is it always intermittent and nondeterministic?

**Comment by BaconPancakes** [comment]
> Is this behavior something that started recently? If so, that would point to a recent code change as the cause.

Yes. This is very recent. I just rolled back to
`Pylance language server 2024.10.1 (pyright version 1.1.382, commit e46efc6b)` and don't see the issue yet (though it could easily be intermittently gone). I'll update if it shows up again.

> If you restart the language server, does the problem go away?

No, I've also tried developer: reload, clear python cache and reload and pylance: clear persisted indices.

> Are you able to repro the problem with the command-line version of pyright?

No dice.

**Comment by failable** [comment]
Yes, it just randomly happens. 

- When I open a project, it's just fine. 
- When I start editing, the errors show. 
- When I save the file, the errors maybe disappear seconds or not. If they disappears, they are shown after seconds.
- repeats...

The issue happens in this recent week.

**Comment by 1hachem** [comment]
thank god I found this issue, this drove me crazy the last three days

**Comment by erictraut** [comment]
> And I've noticed that these kinds of errors can randomly appear

When you say "these kinds of errors", are you referring specifically to the "Coroutine[Any, Any, None] is not awaitable" error, or are there other errors you're seeing? In the absence of a solid repro, I'm trying to establish some pattern, so any clues you can provide are welcome.

**Comment by KimigaiiWuyi** [comment]
> > And I've noticed that these kinds of errors can randomly appear
> 
> When you say "these kinds of errors", are you referring specifically to the "Coroutine[Any, Any, None] is not awaitable" error, or are there other errors you're seeing? In the absence of a solid repro, I'm trying to establish some pattern, so any clues you can provide are welcome.

Basically, I mean `All asynchronous function calls`, i.e. `Coroutine[Any, Any, None] is not awaitable`.
`**randomly**` means, it will mark the wavy lines randomly, or it will mark the wavy lines or unmark the wavy lines after I save them.

**Comment by failable** [comment]
For me, "these kinds of errors" are

-   `"Coroutine[Any, Any, list[TextRegion]]" is incompatible with protocol "Awaitable[_T_co@Awaitable]"    "__await__" is an incompatible type      Type "() -> Generator[Any, Any, list[TextRegion]]" is not assignable to type "() -> Generator[Any, Any, _T_co@Awaitable]"        Function return type "Generator[Any, Any, list[TextRegion]]" is incompatible with type "Generator[Any, Any, _T_co@Awaitable]"          "Generator[Any, Any, list[TextRegion]]" is not assignable to "Generator[Any, Any, _T_co@Awaitable]"`
-  `"Future[list[Drug | BaseException | None]]" is incompatible with protocol "Awaitable[_T_co@Awaitable]"    "__await__" is an incompatible type      Type "() -> Generator[Any, None, list[Drug | BaseException | None]]" is not assignable to type "() -> Generator[Any, Any, _T_co@Awaitable]"        Function return type "Generator[Any, None, list[Drug | BaseException | None]]" is incompatible with type "Generator[Any, Any, _T_co@Awaitable]"          "Generator[Any, None, list[Drug | BaseException | None]]" is not assignable to "Generator[Any, Any, _T_co@Awaitable]"`
-   `"Literal['parse_']" is not assignable to "str"`
-   `"builtins.str" is not assignable to "builtins.str"`

I had 0 of them before, now I have 539 of them.

**Comment by erictraut** [comment]
> I had 0 of them before, now I have 539 of them.

Thanks for the info. When you say "before", do you remember which version you were using?

Has anyone seen the problem with the command-line version of pyright, or is this only with the language server?

I just release version 1.1.385 of pyright. It should be published to pypi in the next 24 hours. The last two errors in your list above are likely addressed by the new version, but I don't think the first two errors will be.

**Comment by failable** [comment]
I don't remember since Zed automatically download  and update `pyright` for me every time. And I'm sure these two issues come at the same time in the recent 10-14 days. I've never seem them before.

**Comment by 1hachem** [comment]
@erictraut this issue still persist event with pyright version `1.1.385` installed

**Comment by demux** [comment]
I have this issue in vscode with Pylance v2024.10.100 (pre-release)

**Comment by AtomicMegaNerd** [comment]
I am seeing this exact same issue in Neovim using pyright installed via Homebrew on my M1 Mac:

```
❯ pyright --version
pyright 1.1.385
```

```
❯ nvim --version
NVIM v0.10.2
Build type: Release
LuaJIT 2.1.1727870382
```

This just started for me about a week or so ago. These errors around async functions seemingly appear and disappear at random.

My project is using Poetry with a local virtualenv at `$PROJECT_ROOT/.venv`. Here is the relevant section of my pyproject.toml file:

```toml
[tool.pyright]
exclude = [".venv", "build", "dist"]
venvPath = "."
venv = ".venv"
```

Many errors like this one:

![Image](https://github.com/user-attachments/assets/640ffa87-f821-4e2b-96ac-9aeb049211de)

**Comment by josiah-lunit** [comment]
Also reporting this problem, specifically after I save or run my linter.  The errors go away after editing the file (but not saving).

```
pyright: 1.1.385
nvim: v0.11.0-nightly+cd8e15e
Build type: Release
LuaJIT 2.1.1713773202
```

Example diagnostic text:
```
Coroutine[Any, Any, T | None] is not awaitable
... is incompatible with protocol "Awaitable[_T_co@Awaitable]"
"__await__" is an incompatible type
Type "() -> Generator[Any, Any, T | None] is not assignable to type () -> Generator[Any, Any, _T_co@Awaitable]
```

**Comment by zivanovicb** [comment]
+1

**Comment by trendels** [comment]
In case it helps, I'm seeing what is probably the same regression, but the reported errors are about assignments of string literals to variables of type `str`:

```
 Diagnostics:
 Argument of type "LiteralString" cannot be assigned to parameter "s" of type "str" in function "write"
   "builtins.str" is not assignable to "builtins.str" [reportArgumentType]
```

This is on Python 3.12. With pyright versions 1.1.384 and  1.1.385. Version 1.1.383 does not show these errors.
The errors are only reported in the editor (neovim with nvim-lspconfig my case), never when running pyright from the command line.
I have not managed to produce a short file to reproduce it, it only happens when I edit a file that is part of a larger codebase. As others have reported, the errors are not shown immediately after opening the file. I have not found a reliable way to trigger it, but for me it seems to require making edits and saving the file before it shows up.

**Comment by erictraut** [comment]
Thanks to everyone for the clues. Unfortunately, I haven't been able to find a repro, and I'm running short on theories at this point. I've tried to repro using a bunch of large open-source code bases that make heavy use of `await` and async IO, but I haven't seen the problem yet.

Are any of you able to repro this with a source base that is publicly accessible in GitHub? 

@trendels, you mentioned that you're seeing an error with `LiteralString`. I don't have any good theories for how that could be caused by the same issue as the "not awaitable" error that others are reporting here. The latter seems to be related to protocol matching, but `LiteralString` and `str` have nothing to do with protocols. Do you have a code sample that reliably repros the issue you're seeing with `LiteralString`?

**Comment by KimigaiiWuyi** [comment]
> I've tried to repro using a bunch of large open-source code bases that make heavy use of await and async IO

I don't think the problem seems to be how big the repo is, as issuse suggests, it can be a problem on a **very short** example, anyway, I'm glad there won't be any problems on top of pyright 1.1.383

**Comment by BaconPancakes** [comment]
I think it's related to `nest_asyncio` in some way.

Here's some self-contained code that will show the error on my machine:
```python
import openai
import nest_asyncio
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from openai.types import chat

nest_asyncio.apply() # pyright: ignore[reportUnknownMemberType]


async def complete(prompt: str) -> str:
    client = openai.AsyncOpenAI(
        base_url="",
    )
    message: chat.chat_completion_message_param.ChatCompletionMessageParam = {
        "role": "user",
        "content": prompt,
    }

    response = await client.chat.completions.create(
        model="MY_MODEL",
        messages=[message],
    )
    return str(response.choices[0].message.content)
```
![Image](https://github.com/user-attachments/assets/33eeb82e-6a4b-4d32-80d9-ef411a359b99)

I'm going to attach a mock repo that contains this code and errors.

I initialized via https://docs.astral.sh/uv/ with `uv sync` and I'm having VSCode point to the generated .venv folder for language server.

 * Pylance language server 2024.10.101 (pyright version 1.1.385, commit 05e8fa01)
 * Operating System: Ubuntu 22.04.4 LTS              
  * Kernel: Linux 5.15.0-101-generic
  * Architecture: x86-64

**Comment by erictraut** [comment]
@BaconPancakes, thanks for taking the time to put together those repro steps. I really appreciate it. It took me a while to get a repro, even with your detailed instructions, but I eventually got it to happen. This allowed me to find and fix the underlying cause.

The underlying bug was introduced a few years ago (relating to special-case handling of `functools.partial`), but the effects of this bug were masked. A recent (seemingly unrelated) change caused this bug to manifest itself. And yes, this bug explains all of the symptoms reported above, including the error involving `LiteralString`.

Thanks everyone for your help in providing clues. Apologies for the regression and the inconvenience this caused.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.386
